R/decostand.R

```diff
@@ -269,3 +311,323 @@
         x[abs(x) < sqrt(.Machine$double.eps)] <- 0
     x
 }
+
+
+OptSpace <- function(x, ROPT=3, NITER=5, TOL=1e-5, verbose=FALSE)
+{
+
+  ## Preprocessing : x     : partially revealed matrix
+  if (is.data.frame(x)) {
+    x <- as.matrix(x)
+  }
+  if (!is.matrix(x)){
+    stop("* OptSpace : an input x should be a matrix")
+  }
+  if (any(is.infinite(x))){
+    stop("* OptSpace : no infinite value in x is allowed")
+  }
+  if (!any(is.na(x))){
+    stop("* OptSpace : there is no unobserved values as NA")
+  }
+  idxna <- (is.na(x))
+  M_E <- array(0, c(nrow(x), ncol(x)))
+  M_E[!idxna] <- x[!idxna]
+  
+  ## Preprocessing : size information
+  n <- nrow(x)
+  m <- ncol(x)
+  
+  ## Preprocessing : other sparse-related concepts
+  nnZ.E <- sum(!idxna)
+  E <- array(0, c(nrow(x), ncol(x))); E[!idxna] <- 1
+  eps <- nnZ.E/sqrt(m*n)
+  
+  ## Preprocessing : ROPT  : implied rank
+  if (is.na(ROPT)){
+    if (verbose){
+      cat("* OptSpace: Guessing an implicit rank.")
+    }
+    r <- min(max(round(.guess_rank(M_E, nnZ.E)), 2), m-1)
+    if (verbose){
+      cat(paste0('* OptSpace: Guessing an implicit rank: Estimated rank : ',r))
+    }
+  } else {
+    r <- round(ROPT)
+    if ((!is.numeric(r)) || (r<1) || (r>m) || (r>n)){
+      stop("* OptSpace: ROPT should be an integer in [1,min(nrow(x),ncol(x))]")
+    }
+  }
+  
+  ## Preprocessing : NITER : maximum number of iterations
+  if ((is.infinite(NITER))||(NITER<=1)||(!is.numeric(NITER))){
+    stop("* OptSpace: invalid NITER number")
+  }
+  NITER <- round(NITER)
+  
+  m0 <- 10000
```

Comment:
what is this constant for?

Reply:
Scaling parameter for gradient calculation. I added a small comment for this internal parameter. References for the method have been provided on the manpage.